<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [63]:
# текст запроса
query_3_1 = f'''select 
                    count(*) as "Количество вакансий"
                from public.vacancies
             '''

In [64]:
# результат запроса
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\1079912922.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [61]:
# текст запроса
query_3_2 = f'''select 
                    count(*) as "Количество работодателей"
                from public.employers
             '''

In [62]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


Количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [59]:
# текст запроса
query_3_3 = f'''select 
                    count(*) as "Количество регионов"
                from public.areas
             '''

In [60]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


Количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [57]:
# текст запроса
query_3_4 = f'''select 
                    count(*) as "Количество сфер деятельности"
                from public.industries
             '''

In [58]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


Количество сфер деятельности
0                           294

***

# выводы по предварительному анализу данных
23501 работодателей предлагают 49197 вакансий в 1362 различных регионах в 294 сферах деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [53]:
# текст запроса
query_4_1 = f'''select 
                    a.name as "Название региона",
                    count(*) as "Количество вакансий"
                from vacancies v
                left join areas a on v.area_id = a.id   
                group by a.name
                order by 2 desc
            ''' #Объеденяем таблицы для отображения названия региона

In [54]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


Название региона  Количество вакансий
0             Москва                 5333
1    Санкт-Петербург                 2851
2              Минск                 2112
3        Новосибирск                 2006
4             Алматы                 1892
..               ...                  ...
764       Тарко-Сале                    1
765    Новоаннинский                    1
766            Бирск                    1
767           Сасово                    1
768             Анна                    1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [65]:
# текст запроса
query_4_2 = f'''select 
                    count(*) as "Вакансий с границами зарплат"
                from vacancies v
                where v.salary_from is not null or v.salary_to is not null
            '''

In [66]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Вакансий с границами зарплат
0                         24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [78]:
# текст запроса
query_4_3 = f'''select 
                    round(avg(v.salary_from)) as "Средняя нижняя граница",
                    round(avg(v.salary_to)) as "Средняя верхняя граница"
                from vacancies v
                
            ''' # Применяется avg для нахождения среднего и round для округления 

In [79]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Средняя нижняя граница  Средняя верхняя граница
0                 71065.0                 110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [103]:
# текст запроса
query_4_4 = f'''select 
                    v.schedule as "Тип рабочего графика",
                    v.employment as "Тип трудоустройства",
                    count(*) as "Кол-во вакансий"
                from vacancies v
                group by 1, 2
                order by 3 desc
            ''' #Группируем по типу и считаем каждое сочетание

In [104]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика  Тип трудоустройства  Кол-во вакансий
0           Полный день     Полная занятость            35367
1      Удаленная работа     Полная занятость             7802
2         Гибкий график     Полная занятость             1593
3      Удаленная работа  Частичная занятость             1312
4        Сменный график     Полная занятость              940
5           Полный день           Стажировка              569
6        Вахтовый метод     Полная занятость              367
7           Полный день  Частичная занятость              347
8         Гибкий график  Частичная занятость              312
9           Полный день     Проектная работа              141
10     Удаленная работа     Проектная работа              133
11        Гибкий график           Стажировка              116
12       Сменный график  Частичная занятость              101
13     Удаленная работа           Стажировка               64
14        Гибкий график     Проектная работа               18
15       Сменный график           Стажировка               12
16       Вахтовый метод     Проектная работа                2
17       Сменный график     Проектная работа                1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [91]:
# текст запроса
query_4_5 = f'''select 
                    v.experience as "Требуемый опыт работы",
                    count(*) as "Количество вакансий"
                from vacancies v
                group by v.experience 
                order by 2 
            '''

In [92]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_13432\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы  Количество вакансий
0           Более 6 лет                 1337
1             Нет опыта                 7197
2         От 3 до 6 лет                14511
3    От 1 года до 3 лет                26152

***

In [35]:
connection.close()
# выводы по детальному анализу вакансий
# Как оказалось, вакансии предложенны не во всех регионах, а в 769 из 1362
# У 24073 вакансий определена хотя бы одна граница зарплаты 
# В начале пути можно рассчитывать на среднюю зп в 70к
# Сменный график, вахтовый метод и проектоная работа требуются меньше всего
# Чем больше опыта, тем меньше вакансий. Также без опыта нелегко найти работу

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [6]:
# текст запроса
query_5_1 = f'''(select 
                    e.name as "Работодатель",
                    count(*) as "Количество вакансий"
                from vacancies v
                left join EMPLOYERS e on v.employer_id = e.id
                group by e.name 
                order by 2 desc
                limit 1)
                
                union all
                
                (select 
                    e.name as "Работодатель",
                    count(*) as "Количество вакансий"
                from vacancies v
                left join EMPLOYERS e on v.employer_id = e.id
                group by e.name 
                order by 2 desc
                limit 1 offset 4) 
            '''
             

In [7]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df


C:\Users\Tolyan\AppData\Local\Temp\ipykernel_4848\2991075485.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Работодатель  Количество вакансий
0         Яндекс                 1933
1  Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [74]:
# текст запроса 
query_5_2 = f'''select
                    a.name "Название региона", 
                    count(distinct e.name) "Кол-во работодателей", 
                    count (v.id) "Кол-во вакансий"
                from areas a
                left join employers e ON e.area=a.id
                left join vacancies v ON v.area_id=a.id 
                where v.id is null
                group by a.id
                order by count(distinct e.name) desc
                limit 1
                '''         

In [75]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_15116\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Название региона  Кол-во работодателей  Кол-во вакансий
0           Россия                   410                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [21]:
# текст запроса
query_5_3 = f'''select
                    e.name as Name, 
                    count(distinct v.area_id) Reg
                from employers e
                left join vacancies v ON v.employer_id = e.id 
                group by e.id
                order by Reg 
                '''     

In [22]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_6008\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  reg
0                    Рокет Ворк    0
1      Джофре Лабортехник, СООО    0
2                       Джетекс    0
3             Алтай Полиметаллы    0
4                Компания Берко    0
...                         ...  ...
23496                 ООО ЕФИН    71
23497    Поляков Денис Иванович   88
23498                Спецремонт  116
23499                Ростелеком  152
23500                    Яндекс  181

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [31]:
# текст запроса
query_5_4 = f'''select 
                    count(e.id) employer
                from employers e
                left join employers_industries ei ON ei.employer_id = e.id 
                left join industries i ON i.id = ei.industry_id 
                where i.name is null
                '''   

In [32]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_6008\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


employer
0      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [75]:
# текст запроса
query_5_5 = f'''select 
                    e.name employer,
                    count(i.id) IND
                from employers e
                left join employers_industries ei ON ei.employer_id = e.id 
                left join industries i ON i.id = ei.industry_id 
                group by employer
                having count(i.id) = 4
                order by employer
                limit 1 offset 2
                '''   

In [76]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_6008\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


employer  ind
0     2ГИС    4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [82]:
# текст запроса
query_5_6 = f'''select 
                    count(e.name) employers
                from employers e
                left join employers_industries ei ON ei.employer_id = e.id 
                left join industries i ON i.id = ei.industry_id 
                where i.name = 'Разработка программного обеспечения'
                '''   

In [83]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_6008\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


employers
0       3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [3]:
# код для получения списка городов-милионников
query_5_7 = f'''(select 
                    a.name ,
                    count(*) cnt
                from VACANCIES v
                left join AREAS a ON a.id = v.area_id 
                where a.name = 'Москва' or a.name ='Санкт-Петербург' or a.name ='Новосибирск' or a.name ='Екатеринбург' or a.name ='Казань'
                                or a.name ='Нижний Новгород' or a.name ='Челябинск' or a.name ='Красноярск' or
                                a.name ='Самара' or a.name ='Уфа' or a.name ='Ростов-на-Дону' or a.name ='Омск' or 
                                a.name ='Краснодар' or a.name ='Воронеж' or a.name ='Пермь' or a.name ='Волгоград'
                group by a.name
                )
                
                union all
                
                (select 
                    'total',
                    count(*) 
                from VACANCIES v
                left join AREAS a ON a.id = v.area_id 
                where a.name = 'Москва' or a.name ='Санкт-Петербург' or a.name ='Новосибирск' or a.name ='Екатеринбург' or a.name ='Казань'
                                or a.name ='Нижний Новгород' or a.name ='Челябинск' or a.name ='Красноярск' or
                                a.name ='Самара' or a.name ='Уфа' or a.name ='Ростов-на-Дону' or a.name ='Омск' or 
                                a.name ='Краснодар' or a.name ='Воронеж' or a.name ='Пермь' or a.name ='Волгоград'
                )
                order by 2
                '''  

In [4]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_15880\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0         Волгоград    456
1              Омск    617
2               Уфа    767
3             Пермь    771
4         Челябинск    786
5        Красноярск    847
6           Воронеж   1063
7    Ростов-на-Дону   1131
8            Самара   1144
9         Краснодар   1301
10           Казань   1415
11  Нижний Новгород   1670
12     Екатеринбург   1698
13      Новосибирск   2006
14  Санкт-Петербург   2851
15           Москва   5333
16            total  23856

In [34]:
# выводы по анализу работодателей
# Яндекс лидирует на рынке
# Работодатели не выставляющие вакансии не привязываются к определённому региону, а оставляют целую страну (Россию)
# Разработка программного обеспечения достаточно востребованная
# Около половины вакансий в городах-миллионниках

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [58]:
# текст запроса
query_6_1 = f'''select 
                    count(*) vac
                from vacancies v
                where name ilike '%данн%' or name ilike '%data%'
                '''   

In [59]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_11424\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


vac
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [88]:
# текст запроса
query_6_2 = f'''select 
                    count(*) vac
                from vacancies v
                where (name ilike '%data scientist%' or name ilike '%data science%' or (name like '%ML%' and name not like '%HTML%')
                        or name ilike '%machine learning%' or name ilike '%машинн%обучен%') and 
                        (name ilike '%junior%' or experience ilike '%нет опыта%' or employment ilike '%стажировка%')
                '''   

In [89]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_11424\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


vac
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [90]:
# текст запроса
query_6_3 = f'''select 
                    count(*) vac
                from vacancies v
                where (name ilike '%data scientist%' or name ilike '%data science%' or (name like '%ML%' and name not like '%HTML%')
                        or name ilike '%machine learning%' or name ilike '%машинн%обучен%')
                        and (key_skills ilike '%sql%' or key_skills ilike '%postgres%')
                '''  

In [91]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_11424\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


vac
0  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [95]:
# текст запроса
query_6_4 = f'''select 
                    count(*) vac
                from vacancies v
                where (name ilike '%data scientist%' or name ilike '%data science%' or (name like '%ML%' and name not like '%HTML%')
                        or name ilike '%machine learning%' or name ilike '%машинн%обучен%')
                        and (key_skills ilike '%python%')
                '''  

In [96]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_11424\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


vac
0  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [27]:
# текст запроса
query_6_5 = f'''with datascience as
                    (select v.name, v.key_skills
                    from vacancies v
                    where name ilike '%data scientist%' or name ilike '%data science%' or (name like '%ML%' and name not like '%HTML%')
                        or name ilike '%machine learning%' or name ilike '%машинн%обучен%')
                select round(avg(length(key_skills)-length(replace(key_skills,CHR(9),''))+1),2) num
                from datascience
                where key_skills is not null
                ''' 

In [28]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_15880\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


num
0  6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [31]:
# текст запроса
query_6_6 = f'''SELECT
                experience,
                ROUND(AVG(COALESCE((salary_to + salary_from)/2, salary_to, salary_from))) average
            FROM public.vacancies
            WHERE (name ilike '%data scientist%' or name ilike '%data science%' or (name like '%ML%' and name not like '%HTML%')
                        or name ilike '%machine learning%' or name ilike '%машинн%обучен%')
                AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                AND (experience IS NOT NULL)
            GROUP BY experience
        '''

In [32]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Tolyan\AppData\Local\Temp\ipykernel_15880\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience   average
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

In [34]:
# выводы по предметному анализу
# Около 3.5% от предложенных вакансий имеет отношение к данным
# Около 50% из них относятся к DS
# Для стажёров DS только 0.1% от всего рынка
# Около 6-7 навыков нужно иметь, чтобы только устроиться на работу 
# Чем больше опыта, тем больше желаемая зп

# Общий вывод по проекту

In [37]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

# Работу легко найти, если ты уже с опытом, но не прям с огромным, потому что тогда не так много мест, которые смогут выплачивать желаемую зп
# Лидирующий работодатель Яндекс, но Сбер, которого здесь нет, начинает тягаться за рынок
# В поисках работы стоит ехать в бОльшие города
# Такие направления как DS и ИИ будут в дальнейшем развиваться, а количество вакансий и их оплачиваемость будут расти